pkg/workflow/controllers/accomplish_watcher.go

```diff
@@ -0,0 +1,147 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+)
+
+type AccomplishWatcher struct {
+	kubeClient    client.Client
+	eventRecorder record.EventRecorder
+	logger        logr.Logger
+}
+
+func NewAccomplishWatcher(kubeClient client.Client, eventRecorder record.EventRecorder, logger logr.Logger) *AccomplishWatcher {
+	return &AccomplishWatcher{kubeClient: kubeClient, eventRecorder: eventRecorder, logger: logger}
+}
+
+func (it *AccomplishWatcher) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.TODO()
+
+	node := v1alpha1.WorkflowNode{}
+	err := it.kubeClient.Get(ctx, request.NamespacedName, &node)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if ConditionEqualsTo(node.Status, v1alpha1.CondiitonAccomplished, corev1.ConditionTrue) ||
+		ConditionEqualsTo(node.Status, v1alpha1.ConditionDeadlineExceed, corev1.ConditionTrue) {
+		owners := node.OwnerReferences
+		// NOOP
+		if len(owners) == 0 {
+			it.logger.V(1).Info("dangling node has no owner", "node", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+
+		// unexpected situation
+		if len(owners) > 1 {
+			it.logger.V(1).Info("node has more than one owner, it will not take any operates", "node", request.NamespacedName, "owners", owners)
+			return reconcile.Result{}, nil
+		}
+
+		if len(owners) == 1 {
+			owner := owners[0]
+			it.logger.V(4).Info("fetch node's owner", "node", request.NamespacedName, "owner", owner)
+			if owner.Kind == GetKindOf(&v1alpha1.WorkflowNode{}) {
+				parentNode := v1alpha1.WorkflowNode{}
+
+				err := it.kubeClient.Get(ctx, types.NamespacedName{
+					Namespace: request.Namespace,
+					Name:      owner.Name,
+				}, &parentNode)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+				if parentNode.Spec.Type == v1alpha1.TypeSerial {
+					err = it.updateParentSerialNode(ctx, node, parentNode)
+					return reconcile.Result{}, err
+				}
+				it.logger.Info("unsupported owner node type", "node type", parentNode.Spec.Type)
+			} else if owner.Kind == GetKindOf(&v1alpha1.Workflow{}) {
+				it.logger.Info("unfinished update for workflow", "kind", owner.Kind)
+			} else {
+				it.logger.Info("unsupported owner type", "kind", owner.Kind)
+			}
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (it *AccomplishWatcher) updateParentSerialNode(ctx context.Context, childNode, parentNode v1alpha1.WorkflowNode) error {
+
+	updateError := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		namespace := parentNode.Namespace
+		name := parentNode.Name
+		parentNode := v1alpha1.WorkflowNode{}
+		err := it.kubeClient.Get(ctx, types.NamespacedName{
+			Namespace: namespace,
+			Name:      name,
+		}, &parentNode)
+
+		if err != nil {
+			return client.IgnoreNotFound(err)
+		}
+
+		// filter out accomplished node
+		var newActiveChildren []corev1.LocalObjectReference
+
+		for _, item := range parentNode.Status.ActiveChildren {
```

Comment:
uhh this is not too good. What about an abstract `filter`

Reply:
Ok, I will extract these lines of codes into a new method, like:
```go
var newActiveChildren []corev1.LocalObjectReference

parentNode.Status.ActiveChildren = filterChildrens(parentNode.Status.ActiveChildren, func(item corev1.LocalObjectReference) bool {
return item.Name != childNode.Name
})
```

Reply:
I thought that there will be something like a haskell `filter` in go but to my surprise there's no such thing. Then it comes to me that there's no generic type in go 🤦 . In fact what I prefer is a
```go
func filter<T>(arr: []T, f: func(T) bool) []T
```
Since there's no such <T> thing, there may have to be some type assert with struct{}.

Reply:
I do not think we will use `struct{}` or `interface{}` if you mean.

> Actually, these codes are copied from kubernetes replicaset controller utils, it also struggling with no generic type.

I think It's ok for now.🙈